### PR TITLE
update db delete

### PR DESCRIPTION
**Pull request by Bento007**
#### Reviewers
**Functional:** @mckinsel 

**Readability:** @tihuan 

---

## Changes
- Add documentation and make recipe to tunnel into database using Bastion
- Add processing_status and tombstone to database
- Modify tests and openapi to support processing status.

### CODE REVIEW

File: backend/database/versions/9a900b8ee3a5_for_deletion_tombstone_and_processing_.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+def upgrade():
+    # Create processing_status enum
+    processing_status_enum = create_enum("processingstatus", ["PENDING", "SUCCESS", "FAILURE"])
+
+    # add processing_status_enum to processing_status column in dataset_process_status table
+    op.add_column("dataset_processing_status", sa.Column("processing_status", processing_status_enum))
+
+    # Add tombstone column to dataset with default false
+    op.add_column("dataset", sa.Column("tombstone", processing_status_enum, default=False))
```
Comment (mckinsel) [inline backend/database/versions/9a900b8ee3a5_for_deletion_tombstone_and_processing_.py:35]: Shouldn't the tombstone columns be a booleans?

File: scripts/populate_db.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ def run_db_stuff(create_schema, recreate_db, populate_data, drop_db):
     # Create schema.
     if create_schema:
         engine = create_engine(CorporaDbConfig().database_uri)
-        if not database_exists(engine.url):
+        if not database_exists(engine.url):  # This never works because the url does not include the database name
```
Comment (mckinsel) [inline scripts/populate_db.py:31]: Can you elaborate a little on that comment?

**Review by mckinsel — CHANGES_REQUESTED** [review]

File: backend/corpora/common/corpora_orm.py [hunk 3]
```diff
@@ -335,6 +337,20 @@ class ConversionStatus(enum.Enum):
...
+    Enumerates the status of processing a dataset.
+
+    PENDING = Processing has not started
+    SUCCESS - Processing succeeded
+    FAILURE - Processing failed
+    """
+
+    PENDING = "pending"
+    SUCCESS = "success"
+    FAILURE = "failure"
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py:351]: These values aren't even used, but for consistency I'll correct

File: backend/corpora/common/entities/collection.py [hunk 4]
```diff
@@ -123,22 +123,28 @@ def to_dict(db_object):
...
-        result.pop("user", None)
-        result.pop("owner", None)
+        for hidden in self._hide:
+            result.pop(hidden, None)
         result["links"] = [
             dict(link_url=link["link_url"], link_name=link["link_name"] or "", link_type=link["link_type"])
             for link in result["links"]
         ]
+
+        result["datasets"] = [ds for ds in result["datasets"] if ds.get("tombtone") is not True]
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/corpora/common/entities/collection.py:143]: The possible values are True, False, or None so your shortened version works.

File: backend/database/versions/9a900b8ee3a5_for_deletion_tombstone_and_processing_.py (hunk 1 above)
Comment (Bento007) [reply backend/database/versions/9a900b8ee3a5_for_deletion_tombstone_and_processing_.py:35]: yes!

File: backend/database/README.md [hunk 5]
```diff
@@ -5,15 +5,33 @@ General information about Alembic migrations can be found [here](https://alembic
...
+1. Makes changes to [corpora_orm.py](../corpora/common/corpora_orm.py)
 1. Set DEPLOYMENT_STAGE=dev, CORPORA_LOCAL_DEV=1
 1. Run `make db/connect`
 1. Follow [How to perform a database migration](#how_to_perform_a_database_migration) instructions except use `make db/new_migration_auto MESSAGE="purpose_of_migration"`. See [What does Autogenerate Detect (and what does it not detect?)](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#what-does-autogenerate-detect-and-what-does-it-not-detect).
+
+## Connect to Remote RDS
+Enable local connection to the private RDS instance:
+
+```Shell
+cd ./data-poral/backend
```
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/database/README.md:26]: fixed

File: backend/config/corpora-api.yml [hunk 6]
```diff
@@ -329,6 +329,10 @@ paths:
                 properties:
                   dataset_uuid:
                     $ref: "#/components/schemas/dataset_uuid"
+                  processing_status:
+                    type: string
+                    enum:
+                      [ PENDING, SUCCESS,FAILURE]
```
Comment (tihuan) [inline backend/config/corpora-api.yml:335]: nit: uneven spacing! e.g., `[ PENDING, SUCCESS,FAILURE]` to `[PENDING, SUCCESS, FAILURE]`

I'll add a commit to run `prettier` on yaml 👌

File: backend/corpora/common/corpora_orm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tihuan) [reply backend/corpora/common/corpora_orm.py:351]: Should we remove `processing_status` related code change from this PR? Given that we already have a separate PR for that [here](https://github.com/chanzuckerberg/corpora-data-portal/pull/867/files)

Thank you!

File: backend/database/README.md [hunk 7]
```diff
@@ -5,15 +5,30 @@ General information about Alembic migrations can be found [here](https://alembic
...
-5. If no errors are outputted, complete the migration by running `DEPLOYMENT_STAGE=dev make db/migrate`. Ensure that tunneling is set up and `CORPORA_LOCAL_DEV` is set to true.
+1. Run `make db/new_migration MESSAGE="purpose_of_migration"` where `purpose_of_migration` is a short phrase describing why the database migration is occurring.
+1. Head over to the file that was just created. The path will be something like `backend/database/versions/xxxxxxxxxxxx_purpose_of_migration.py`. Edit the `upgrade()` and `downgrade()` functions such that `upgrade()` contains the commands to perform the migration you would like and `downgrade()` contains the commands to undo it.
+1. Enable local connection to the private `dev` RDS instance through Bastion: `make db/tunnel`. See [Connect to Remote RDS](#connect_to_remote_rds) for more details.
```
Comment (tihuan) [inline backend/database/README.md:10]: super nice documentation! Thank you!

File: backend/database/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (tihuan) [reply backend/database/README.md:26]: maybe the commit is not pushed yet, but I still see `cd ./data-poral/backend` 😆

File: backend/corpora/common/entities/collection.py [hunk 8]
```diff
@@ -123,22 +123,28 @@ def to_dict(db_object):
 
         return results
 
+    _hide = ["user", "owner", "tombstone"]
```
[the remark this answers is not in the document]
Comment (tihuan) [reply backend/corpora/common/entities/collection.py:126]: I'm fine either way, but if we're going with this constant, [Python style guide](https://google.github.io/styleguide/pyguide.html#3164-guidelines-derived-from-guidos-recommendations) seems to say it should be `ALL_CAPS`?

E.g., `_hide` to `_HIDE`

**Review by tihuan — APPROVED** [review]
LGTM with reservation for pulling out `processing_status` related code? Thank you!

File: scripts/populate_db.py (hunk 2 above)
Comment (Bento007) [reply scripts/populate_db.py:31]: I took out the comment and filed a bug https://github.com/chanzuckerberg/corpora-data-portal/issues/870

File: backend/corpora/common/corpora_orm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bento007) [reply backend/corpora/common/corpora_orm.py]: that PR is based off this PR

**Review by mckinsel — APPROVED** [review]
👍

**Pull request merged**